File: 08_Codebases/iOS_App/supabase/functions/_shared/meta_capi.ts
/**
 * Meta Conversions API (CAPI) — shared Deno helper for Edge Functions.
 *
 * Used by:
 *   - auth-hook                 → CompleteRegistration after signUp
 *   - meta-capi-event           → generic relay for DB webhooks
 *   - stripe-webhook            → Purchase on payment_intent.succeeded
 *   - save-early-access-lead    → Lead on the pre-launch funnel
 *
 * All four import { sendCapiEvent, buildEventId, extractClientIp } from
 * here. The module is intentionally framework-free (no Supabase SDK,
 * no Stripe SDK) so the same surface can be added to any future Edge
 * Function without dragging in deps.
 *
 * Why server-side at all (recap):
 *   Browser pixel events get blocked by iOS 14+ ATT (~40% loss), ad
 *   blockers (~25% of users), and Safari ITP cookie wipes. CAPI sends
 *   the same conversion from a datacenter IP with a token the browser
 *   never sees — none of those layers can intercept it.
 *
 * Required secrets (Supabase project secrets):
 *   META_PIXEL_ID=1445760663897743
 *   META_ACCESS_TOKEN=<System User token from Events Manager, doesn't expire>
 *   META_API_VERSION=v22.0
 *
 * Optional during QA — events flow to Events Manager → Test Events tab
 * instead of production:
 *   META_TEST_EVENT_CODE=TESTxxxxx
 *
 * Deduplication with the browser pixel: pass the SAME `event_id` to
 * both `fbq('track', name, params, { eventID })` AND
 * `sendCapiEvent({ event_id })`. Meta keeps one copy per
 * (event_name, event_id) within a 7-day window. Build the id with
 * `buildEventId(eventType, entityId)` on both sides so the value
 * matches without any explicit coordination.
 */

const META_PIXEL_ID = Deno.env.get("META_PIXEL_ID") ?? "1445760663897743";
const META_ACCESS_TOKEN = Deno.env.get("META_ACCESS_TOKEN") ?? "";
const META_API_VERSION = Deno.env.get("META_API_VERSION") ?? "v22.0";
const META_TEST_EVENT_CODE = Deno.env.get("META_TEST_EVENT_CODE") || undefined;

const META_CAPI_URL =
  `https://graph.facebook.com/${META_API_VERSION}/${META_PIXEL_ID}/events`;

// 5s mirrors the reference Python module. Meta P95 is well under 1s; the
// extra headroom absorbs regional latency without holding a caller (e.g.
// the Stripe webhook) past Stripe's own 10s retry budget.
const HTTP_TIMEOUT_MS = 5000;

export type ActionSource =
  | "website"
  | "app"
  | "physical_store"
  | "system_generated"
  | "email"
  | "chat";

export interface CapiUserData {
  email?: string;
  /** E.164 without "+", e.g. "393331234567". Non-digits stripped before hash. */
  phone?: string;
  /** Internal user id (Supabase auth.users.id). Hashed before send. */
  externalId?: string | number;
  /** Real client IP — extract via extractClientIp(req). */
  clientIp?: string;
  /** Browser UA — req.headers.get("user-agent"). */
  clientUserAgent?: string;
  fbp?: string;
  fbc?: string;
}

export interface SendCapiEventArgs extends CapiUserData {
  eventName: string;
  eventId: string;
  value?: number;
  currency?: string;
  customData?: Record<string, unknown>;
  actionSource?: ActionSource;
  eventSourceUrl?: string;
}

/**
 * Deterministic event_id shared with the browser pixel for dedup.
 *
 *   buildEventId("registration", user.id)        → "registration_<uuid>"
 *   buildEventId("purchase", booking.id)         → "purchase_<uuid>"
 *   buildEventId("lead", emailHash + bucket)     → "lead_<short>_<ts>"
 *
 * Use the same key on both sides (frontend pixel + this server) so Meta
 * collapses the pair into one event with the union of their signals.
 */
export function buildEventId(
  eventType: string,
  entityId: string | number,
): string {
  return `${eventType}_${entityId}`;
}

/**
 * Pull the real client IP from a Request, honouring reverse-proxy
 * headers Supabase Edge Network sets. Matches the reference Python
 * `extract_client_ip` order so behavior is consistent across surfaces.
 */
export function extractClientIp(req: Request): string | undefined {
  const xff = req.headers.get("x-forwarded-for");
  if (xff) return xff.split(",")[0]?.trim() || undefined;
  const xri = req.headers.get("x-real-ip");
  if (xri) return xri.trim();
  return undefined;
}

async function sha256Hex(value: string): Promise<string> {
  const data = new TextEncoder().encode(value.trim().toLowerCase());
  const buf = await crypto.subtle.digest("SHA-256", data);
  return Array.from(new Uint8Array(buf))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

async function buildUserData(d: CapiUserData): Promise<Record<string, unknown>> {
  const ud: Record<string, unknown> = {};
  if (d.email) ud.em = [await sha256Hex(d.email)];
  if (d.phone) {
    const digits = d.phone.replace(/\D/g, "");
    if (digits) ud.ph = [await sha256Hex(digits)];
  }
  if (
    d.externalId !== undefined && d.externalId !== null && d.externalId !== ""
  ) {
    ud.external_id = [await sha256Hex(String(d.externalId))];
  }
  if (d.clientIp) ud.client_ip_address = d.clientIp;
  if (d.clientUserAgent) ud.client_user_agent = d.clientUserAgent;
  if (d.fbp) ud.fbp = d.fbp;
  if (d.fbc) ud.fbc = d.fbc;
  return ud;
}

/**
 * Send one event to Meta's Conversions API.
 *
 * Returns the parsed JSON response on success, null on any failure —
 * by design this never throws and never blocks the caller, because the
 * caller is a conversion-critical path (webhook, signup confirmation)
 * where a dropped marketing event is strictly preferable to a broken
 * user flow.
 *
 * Fail-silent matrix:
 *   - Meta 4xx/5xx           → log + null. Don't retry; next genuine
 *                              conversion re-attempts and the lost
 *                              event is basis-points attribution loss,
 *                              not a user-facing bug.
 *   - Network timeout (>5s)  → log + null. Same reasoning.
 *   - Missing access token   → log + null. Lets staging boot without
 *                              the prod secret configured.
 *
 * Payload shape is byte-for-byte equivalent to the reference Python
 * module so Meta's `events_received` counter, Match Quality scoring,
 * and dedup against the browser pixel all behave identically.
 */
export async function sendCapiEvent(
  args: SendCapiEventArgs,
): Promise<Record<string, unknown> | null> {
  if (!META_ACCESS_TOKEN) {
    console.warn(
      `[META CAPI] META_ACCESS_TOKEN not configured, skipping ${args.eventName}`,
    );
    return null;
  }

  const userData = await buildUserData(args);
  const eventPayload: Record<string, unknown> = {
    event_name: args.eventName,
    event_time: Math.floor(Date.now() / 1000),
    event_id: args.eventId,
    action_source: args.actionSource ?? "website",
    user_data: userData,
  };
  if (args.eventSourceUrl) eventPayload.event_source_url = args.eventSourceUrl;

  const customData: Record<string, unknown> = {};
  if (args.eventName === "Purchase" && typeof args.value === "number") {
    customData.value = Math.round(args.value * 100) / 100;
    customData.currency = (args.currency ?? "EUR").toUpperCase();
  }
  if (args.customData) Object.assign(customData, args.customData);
  if (Object.keys(customData).length > 0) {
    eventPayload.custom_data = customData;
  }

  const body: Record<string, unknown> = {
    data: [eventPayload],
    access_token: META_ACCESS_TOKEN,
  };
  if (META_TEST_EVENT_CODE) body.test_event_code = META_TEST_EVENT_CODE;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), HTTP_TIMEOUT_MS);

  try {
    const response = await fetch(META_CAPI_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
      signal: controller.signal,
    });

    const text = await response.text();
    let data: Record<string, unknown> = {};
    try {
      data = text ? JSON.parse(text) as Record<string, unknown> : {};
    } catch {
      // Meta occasionally returns plain text on 5xx; preserve in log
      // so debugging doesn't require replaying the request.
    }

    if (!response.ok) {
      console.error(
        `[META CAPI] HTTP ${response.status} on ${args.eventName}: ${
          text.slice(0, 500)
        }`,
      );
      return null;
    }

    const received = Number(data.events_received ?? 0);
    if (received > 0) {
      console.log(
        `[META CAPI] ${args.eventName} OK (id=${args.eventId}, received=${received}, fbtrace=${
          data.fbtrace_id ?? ""
        })`,
      );
    } else {
      console.warn(
        `[META CAPI] ${args.eventName} sent but events_received=0 (id=${args.eventId}, resp=${
          JSON.stringify(data).slice(0, 300)
        })`,
      );
    }
    return data;
  } catch (err) {
    if (err instanceof Error && err.name === "AbortError") {
      console.warn(
        `[META CAPI] timeout on ${args.eventName} (id=${args.eventId})`,
      );
    } else {
      console.error(
        `[META CAPI] unhandled error on ${args.eventName}: ${
          err instanceof Error ? err.message : String(err)
        }`,
      );
    }
    return null;
  } finally {
    clearTimeout(timer);
  }
}

File: Temporary_/backend/app/services/meta_capi.py
"""
Meta Conversions API (CAPI) integration for Holistic Unity.

Sends server-side events to Meta Pixel for accurate attribution
immune to ad blockers and iOS tracking restrictions.

Events tracked:
  - CompleteRegistration: cliente crea account
  - Lead: pratico completa onboarding/registrazione
  - Purchase: sessione prenotata e pagata (Stripe webhook)

Pixel ID: 1445760663897743 (Holistic Unity_Dataset)

Usage:
  from app.services.meta_capi import send_capi_event, build_event_id, extract_client_ip

  await send_capi_event(
      event_name="CompleteRegistration",
      event_id=build_event_id("registration", user.id),
      email=user.email,
      client_ip=extract_client_ip(request),
      client_user_agent=request.headers.get("user-agent"),
      action_source="website",  # o "app" se dall'app iOS
      event_source_url=str(request.url),
  )

Setup .env:
  META_PIXEL_ID=1445760663897743
  META_ACCESS_TOKEN=<system user token>
  META_API_VERSION=v22.0
  # META_TEST_EVENT_CODE=TEST12345  # opzionale, solo durante testing

Deduplicazione pixel-web ↔ CAPI:
  Usa SEMPRE lo stesso event_id sia nel pixel JS che nel CAPI.
  Meta deduplica automaticamente per (event_name, event_id) entro 7 giorni.
"""
import hashlib
import logging
import os
import time
from typing import Optional

import httpx

logger = logging.getLogger(__name__)

# ── Configuration ────────────────────────────────────────────────────────────
META_PIXEL_ID = os.getenv("META_PIXEL_ID", "1445760663897743")
META_ACCESS_TOKEN = os.getenv("META_ACCESS_TOKEN")
META_API_VERSION = os.getenv("META_API_VERSION", "v22.0")
# Opzionale: setta solo durante testing per vedere eventi in Events Manager → Test Events
META_TEST_EVENT_CODE: Optional[str] = os.getenv("META_TEST_EVENT_CODE") or None

META_CAPI_URL = f"https://graph.facebook.com/{META_API_VERSION}/{META_PIXEL_ID}/events"
HTTP_TIMEOUT_SEC = 5.0

# ── Shared async HTTP client (connection pool) ──────────────────────────────
_client: Optional[httpx.AsyncClient] = None


def _get_client() -> httpx.AsyncClient:
    """Lazily create a shared httpx client. Reuses connection pool."""
    global _client
    if _client is None:
        _client = httpx.AsyncClient(timeout=HTTP_TIMEOUT_SEC)
    return _client


async def close_capi_client() -> None:
    """Call this on FastAPI shutdown (app.on_event('shutdown'))."""
    global _client
    if _client is not None:
        await _client.aclose()
        _client = None


# ── Helpers ──────────────────────────────────────────────────────────────────
def _hash(value: str) -> str:
    """SHA-256 hash normalizzato (lowercase + trim), come richiesto da Meta CAPI."""
    return hashlib.sha256(value.strip().lower().encode()).hexdigest()


def build_event_id(event_type: str, entity_id) -> str:
    """
    Costruisce un event_id deterministico per la deduplicazione pixel↔CAPI.

    Examples:
        build_event_id("registration", user.id)  → "registration_42"
        build_event_id("lead", practitioner.id) → "lead_17"
        build_event_id("purchase", session_id)  → "purchase_cs_test_a1B2c3D4"
    """
    return f"{event_type}_{entity_id}"


def extract_client_ip(request) -> Optional[str]:
    """
    Estrae l'IP del client da una Request FastAPI, gestendo reverse proxy.

    Order of preference:
      1. X-Forwarded-For (Nginx, Cloudflare): primo IP della catena
      2. X-Real-IP (Nginx alt)
      3. request.client.host (direct)
    """
    xff = request.headers.get("x-forwarded-for")
    if xff:
        return xff.split(",")[0].strip()
    xri = request.headers.get("x-real-ip")
    if xri:
        return xri.strip()
    if request.client and request.client.host:
        return request.client.host
    return None


# ── Main API ─────────────────────────────────────────────────────────────────
async def send_capi_event(
    event_name: str,
    event_id: str,
    *,
    email: Optional[str] = None,
    phone: Optional[str] = None,  # formato E.164 senza '+', es. "393331234567"
    external_id: Optional[str] = None,  # user_id interno (verrà hashato)
    client_ip: Optional[str] = None,
    client_user_agent: Optional[str] = None,
    fbp: Optional[str] = None,  # cookie _fbp del browser (se disponibile)
    fbc: Optional[str] = None,  # cookie _fbc del browser (se disponibile)
    value: Optional[float] = None,
    currency: str = "EUR",
    custom_data: Optional[dict] = None,
    action_source: str = "website",
    event_source_url: Optional[str] = None,
) -> Optional[dict]:
    """
    Invia un evento server-side al Meta Conversions API.

    Args:
        event_name: Lead | CompleteRegistration | Purchase | AddToCart | InitiateCheckout | etc.
        event_id: ID univoco per deduplicazione pixel-web ↔ CAPI
        email: email utente (hashata SHA-256)
        phone: numero di telefono in formato E.164 senza '+' (hashato)
        external_id: ID utente interno (hashato — alza match rate ~10-15%)
        client_ip: IP del client (raccomandato per match rate)
        client_user_agent: User-Agent del client (raccomandato per match rate)
        fbp: cookie '_fbp' del browser, se passato dal frontend
        fbc: cookie '_fbc' del browser, se passato dal frontend
        value: importo (solo per Purchase, in account currency)
        currency: ISO 4217, default EUR
        custom_data: dict aggiuntivi (content_name, content_ids, num_items, predicted_ltv, ...)
        action_source: 'website' | 'app' | 'physical_store' | 'system_generated' | 'email' | 'chat'
        event_source_url: URL della pagina sorgente (per action_source='website')

    Returns:
        dict response Meta se OK, None se errore. Non blocca MAI il flusso del chiamante
        (logga e ritorna None in caso di errore).
    """
    if not META_ACCESS_TOKEN:
        logger.warning("[META CAPI] META_ACCESS_TOKEN non configurato, skip %s", event_name)
        return None

    user_data: dict = {}
    if email:
        user_data["em"] = [_hash(email)]
    if phone:
        digits = "".join(c for c in phone if c.isdigit())
        if digits:
            user_data["ph"] = [_hash(digits)]
    if external_id is not None:
        user_data["external_id"] = [_hash(str(external_id))]
    if client_ip:
        user_data["client_ip_address"] = client_ip
    if client_user_agent:
        user_data["client_user_agent"] = client_user_agent
    if fbp:
        user_data["fbp"] = fbp
    if fbc:
        user_data["fbc"] = fbc

    event_payload: dict = {
        "event_name": event_name,
        "event_time": int(time.time()),
        "event_id": event_id,
        "action_source": action_source,
        "user_data": user_data,
    }
    if event_source_url:
        event_payload["event_source_url"] = event_source_url

    cd: dict = {}
    if event_name == "Purchase" and value is not None:
        cd["value"] = round(float(value), 2)
        cd["currency"] = currency
    if custom_data:
        cd.update(custom_data)
    if cd:
        event_payload["custom_data"] = cd

    payload: dict = {
        "data": [event_payload],
        "access_token": META_ACCESS_TOKEN,
    }
    if META_TEST_EVENT_CODE:
        payload["test_event_code"] = META_TEST_EVENT_CODE

    try:
        response = await _get_client().post(META_CAPI_URL, json=payload)
        response.raise_for_status()
        data = response.json()
        received = data.get("events_received", 0)
        if received > 0:
            logger.info(
                "[META CAPI] %s OK (id=%s, received=%s, fbtrace=%s)",
                event_name, event_id, received, data.get("fbtrace_id"),
            )
        else:
            logger.warning(
                "[META CAPI] %s sent ma events_received=0 (id=%s, resp=%s)",
                event_name, event_id, data,
            )
        return data
    except httpx.HTTPStatusError as e:
        logger.error(
            "[META CAPI] HTTP %s on %s: %s",
            e.response.status_code, event_name, e.response.text[:500],
        )
    except httpx.TimeoutException:
        logger.warning("[META CAPI] timeout on %s (id=%s)", event_name, event_id)
    except Exception as e:  # pragma: no cover
        logger.exception("[META CAPI] unhandled error on %s: %s", event_name, e)

    return None  # fail silently, do not block caller
